migrations/2025-06-04-103308_multi-community/up.sql

```diff
@@ -0,0 +1,49 @@
+CREATE TABLE multi_community (
+    id serial PRIMARY KEY,
+    creator_id int NOT NULL REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE,
+    instance_id int NOT NULL REFERENCES instance ON UPDATE CASCADE ON DELETE CASCADE,
+    name varchar(255) NOT NULL,
+    title varchar(255),
+    description varchar(255),
+    local bool NOT NULL DEFAULT TRUE,
+    deleted bool NOT NULL DEFAULT FALSE,
+    ap_id text UNIQUE NOT NULL DEFAULT generate_unique_changeme (),
+    public_key text NOT NULL,
+    private_key text,
+    inbox_url text NOT NULL DEFAULT generate_unique_changeme (),
+    last_refreshed_at timestamptz NOT NULL DEFAULT now(),
+    following_url text NOT NULL DEFAULT generate_unique_changeme (),
+    published timestamptz NOT NULL DEFAULT now(),
+    updated timestamptz
+);
+
+CREATE TABLE multi_community_entry (
+    multi_community_id int NOT NULL REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE,
+    community_id int NOT NULL REFERENCES community ON UPDATE CASCADE ON DELETE CASCADE,
+    PRIMARY KEY (multi_community_id, community_id)
+);
+
+CREATE TABLE multi_community_follow (
+    multi_community_id int NOT NULL REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE,
+    person_id int NOT NULL REFERENCES person ON UPDATE CASCADE ON DELETE CASCADE,
+    follow_state community_follower_state NOT NULL,
+    PRIMARY KEY (multi_community_id, person_id)
+);
+
+ALTER TABLE local_site
+    ADD COLUMN suggested_communities int REFERENCES multi_community ON UPDATE CASCADE ON DELETE CASCADE;
+
+ALTER TYPE listing_type_enum
+    ADD VALUE 'Suggested';
+
+ALTER TABLE community_actions
+    ADD COLUMN is_multi_community_follow bool;
```

Comment:
Okay I see this now. But the scary thing here is that this table has a primary key of `person_id, community_id`.

If a community is part of a multi-sub, then you do an ordinary subscribe, it needs to get *upgraded* by setting `is_multi_community_follow` to false. 

If you unfollow, it should only unfollow rows where `is_multi_community` is false. 

This is a bit complicated to try to put both of these in one table, and really it could be solved easier by having a fake multi_community follower user handle all these, rather than adding this column.

Reply:
Another solution is to only add a `community_actions.multi_community_follows` column, which counts the number of multi-communities that both contain this community and are followed by this user. This allows an approach that's very similar to reference counting. Following a multi-community increments, and unfollowing decrements. 0 must be stored as null because of how uplete works.

Subscribed-only filters would become `community_actions::followed.is_not_null().or(community_actions::multi_community_follows.is_not_null())`.

Reply:
Im implementing a dummy user to follow these communities and getting rid of this column.